### PR TITLE
Publish/delete Agent Group Mutation

### PR DESCRIPTION
**Pull request by elyousfi5**

### CODE REVIEW

File: src/ostorlab/serve_app/common.py [hunk 1]
```diff
@@ -27,6 +29,48 @@ class RiskRatingEnum(graphene.Enum):
     INFO = "Info"
 
 
+class Bytes(scalars.Scalar):
+    """
+    The `Bytes` scalar type represents binary data in a bytes format.
+    """
+
+    @staticmethod
+    def coerce_bytes(value: Union[str, bytes, memoryview, list]) -> bytes:
```
Comment (mohsinenar) [inline src/ostorlab/serve_app/common.py:38]: DocString!

File: src/ostorlab/serve_app/types.py [hunk 2]
```diff
@@ -342,3 +342,51 @@ class OxoScansType(graphene.ObjectType):
     """Graphene object type for a list of scans."""
 
     scans = graphene.List(OxoScanType, required=True)
+
+
+class AgentArgumentInputType(graphene.InputObjectType):
```
Comment (mohsinenar) [inline src/ostorlab/serve_app/types.py:347]: Missing DocString!

File: src/ostorlab/runtimes/local/models/models.py [hunk 3]
```diff
@@ -5,8 +5,8 @@
 import json
 import logging
 import pathlib
-from typing import Any, Dict, List, Optional
 import types
+from typing import Any, Dict, List, Optional
```
Comment (oussamaessaji) [inline src/ostorlab/runtimes/local/models/models.py:9]:
```suggestion
from typing import Any
```

Comment (3asm) [reply src/ostorlab/runtimes/local/models/models.py:9]: Python 3.9 support. We can't  switch yet.

File: src/ostorlab/runtimes/local/models/models.py [hunk 4]
```diff
@@ -386,12 +384,13 @@ class AgentGroup(Base):
     )
 
     @staticmethod
-    def create(name: str, description: str) -> "AgentGroup":
+    def create(name: str, description: str, agents: Any) -> "AgentGroup":
         """Persist the agent group in the database.
 
         Args:
-            key: Agent group key.
+            name: Agent group key.
```
Comment (3asm) [inline src/ostorlab/runtimes/local/models/models.py:391]: Why this change?

File: src/ostorlab/serve_app/common.py [hunk 5]
```diff
@@ -27,6 +29,57 @@ class RiskRatingEnum(graphene.Enum):
     INFO = "Info"
 
 
+class Bytes(scalars.Scalar):
```
Comment (3asm) [inline src/ostorlab/serve_app/common.py:32]: Where do we need that? This is for ubjson.

File: src/ostorlab/serve_app/oxo.py [hunk 6]
```diff
@@ -177,8 +177,80 @@ def mutate(
             return DeleteScanMutation(result=True)
 
 
+class PublishAgentGroupMutation(graphene.Mutation):
```
Comment (3asm) [inline src/ostorlab/serve_app/oxo.py:180]: Create

**Review by 3asm — CHANGES_REQUESTED** [review]

File: src/ostorlab/runtimes/local/models/models.py (hunk 4 above)
Comment (ErebusZ) [reply src/ostorlab/runtimes/local/models/models.py:391]: its not a change, its just a correction

File: src/ostorlab/serve_app/oxo.py [hunk 7]
```diff
@@ -177,8 +177,80 @@ def mutate(
...
+            DeleteAgentGroupMutation: Delete agent group mutation.
+        """
+        with models.Database() as session:
+            agent_group_query = session.query(models.AgentGroup).filter_by(
+                id=agent_group_id
+            )
+            if agent_group_query.count() == 0:
+                raise graphql.GraphQLError("AgentGroup not found.")
+            agent_group_query.delete()
+            session.commit()
```
Comment (deadly-panda) [inline src/ostorlab/serve_app/oxo.py:242]: Do we need to delete all the agents related to this agent group?
I think we should've added `cascade delete` on the models..

File: src/ostorlab/serve_app/types.py [hunk 8]
```diff
@@ -342,3 +342,55 @@ class OxoScansType(graphene.ObjectType):
...
+    """Input object type for an agent argument."""
+
+    name = graphene.String(required=True)
+    type = graphene.String(required=True)
+    description = graphene.String(required=False)
+    value = common.Bytes(required=False)
+
+
+class AgentGroupAgentCreateInputType(graphene.InputObjectType):
```
Comment (deadly-panda) [inline src/ostorlab/serve_app/types.py:356]:
```suggestion
class AgentGroupAgentInputType(graphene.InputObjectType):
```

File: src/ostorlab/serve_app/types.py [hunk 9]
```diff
@@ -342,3 +342,55 @@ class OxoScansType(graphene.ObjectType):
...
+        model = models.AgentGroup
+
+        only_fields = (
+            "id",
+            "name",
+            "description",
+            "created_time",
+        )
+
+    def resolve_key(self: models.AgentGroup, info: graphql_base.ResolveInfo) -> str:
```
Comment (deadly-panda) [inline src/ostorlab/serve_app/types.py:386]: Shouldn't we also resolve the list of agents or no need to?

File: tests/serve_app/oxo_test.py [hunk 10]
```diff
@@ -459,3 +459,100 @@ def testDeleteScanMutation_whenScanDoesNotExist_returnErrorMessage(
 
     assert response.status_code == 200, response.get_json()
     assert response.get_json()["errors"][0]["message"] == "Scan not found."
+
+
+def testPublishAgentGroupMutation_always_shouldPublishAgentGroup(
+    client: testing.FlaskClient,
+) -> None:
+    """Ensure push agent group create an agent group."""
```
Comment (deadly-panda) [inline tests/serve_app/oxo_test.py:467]:
```suggestion
    """Ensure the publish agent group mutation creates an agent group."""
```

File: tests/conftest.py [hunk 11]
```diff
@@ -758,3 +758,17 @@ def android_scan(clean_db: None) -> None:
...
+
+
+@pytest.fixture
+def agent_group() -> models.AgentGroup:
+    """Create dummy agent group."""
+    with models.Database() as session:
+        agent_group = models.AgentGroup(
+            name="Agent Group 1",
+            description="Agent Group 1 description",
+            created_time=datetime.datetime.now(),
```
Comment (deadly-panda) [inline tests/conftest.py:770]: agents

File: src/ostorlab/runtimes/local/models/models.py (hunk 4 above)
Comment (elyousfi5) [reply src/ostorlab/runtimes/local/models/models.py:391]: just to adapt with RE, we have the name in the model, and the key will be resolved in the graphene type

File: src/ostorlab/serve_app/oxo.py (hunk 6 above)
Comment (elyousfi5) [reply src/ostorlab/serve_app/oxo.py:180]: i named it `PublishAgentGroupMutation` to align with RE !

File: src/ostorlab/serve_app/oxo.py (hunk 7 above)
Comment (elyousfi5) [reply src/ostorlab/serve_app/oxo.py:242]: why ? agents and agent group have many2many relation, so an agent used in this agent group might be used in others, so no need to delete.

File: src/ostorlab/serve_app/types.py (hunk 8 above)
Comment (elyousfi5) [reply src/ostorlab/serve_app/types.py:356]: this is named `AgentGroupAgentCreateInputType` to align with RE

File: src/ostorlab/serve_app/types.py (hunk 9 above)
Comment (elyousfi5) [reply src/ostorlab/serve_app/types.py:386]: @3asm shall we ? because we dont do that in RE spec

File: tests/conftest.py (hunk 11 above)
Comment (elyousfi5) [reply tests/conftest.py:770]: why ?, this is to test the deletion of agent_group, the mapping between agents and agent group is tested on the graphql api tests and the models tests

File: src/ostorlab/serve_app/common.py (hunk 5 above)
Comment (elyousfi5) [reply src/ostorlab/serve_app/common.py:32]: it is used for the value of AgentArgumentInputType

File: src/ostorlab/serve_app/oxo.py (hunk 6 above)
Comment (3asm) [reply src/ostorlab/serve_app/oxo.py:180]: Ok.

File: src/ostorlab/serve_app/common.py (hunk 5 above)
Comment (3asm) [reply src/ostorlab/serve_app/common.py:32]: This means we need `ubjson` on the OXO API.

Comment (elyousfi5) [reply src/ostorlab/serve_app/common.py:32]: @3asm added, please check

File: src/ostorlab/runtimes/local/models/models.py (hunk 4 above)
Comment (3asm) [inline src/ostorlab/runtimes/local/models/models.py:391]: We should use the key since we don't have the notion of organization.

File: src/ostorlab/serve_app/app.py [hunk 12]
```diff
@@ -9,8 +10,25 @@ def create_app(path: str = "/graphql", **kwargs) -> flask.Flask:
...
+class CustomUBJSONFileUploadGraphQLView(graphene_upload_flask.FileUploadGraphQLView):
+    """Handles application/ubjson content type in flask views"""
+
+    def parse_body(self):
+        """Handle application/ubjson content type"""
+
+        content_type = flask.request.mimetype
+        if content_type == "application/ubjson":
+            return self.ubjson_decode(flask.request)
+        return super().parse_body()
```
Comment (3asm) [inline src/ostorlab/serve_app/app.py:29]: Missing the response part that can also be ubjson.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: src/ostorlab/runtimes/local/models/models.py (hunk 4 above)
Comment (elyousfi5) [reply src/ostorlab/runtimes/local/models/models.py:391]: as discussed, there will be name and key both

File: src/ostorlab/serve_app/app.py [hunk 13]
```diff
@@ -30,3 +34,79 @@ def authenticate() -> Optional[tuple[flask.Response, int]]:
...
+                content_type = "application/ubjson"
+            else:
+                result, status_code = graphql_server.encode_execution_results(
+                    execution_results,
+                    is_batch=isinstance(data, list),
+                    format_error=self.format_error,
+                    encode=functools.partial(self.encode, pretty=pretty),
+                )
+
+            if show_graphiql:
```
Comment (PiranhaSa) [inline src/ostorlab/serve_app/app.py:101]: explicit please

File: src/ostorlab/serve_app/common.py [hunk 14]
```diff
@@ -40,6 +42,57 @@ class RiskRatingEnum(graphene.Enum):
...
+        elif isinstance(value, list):
+            return bytes(value)
+        else:
+            raise NotImplementedError(f"Bytes scalar coerce error from {type(value)}")
+
+    serialize = coerce_bytes
+    parse_value = coerce_bytes
+
+    @staticmethod
+    def parse_literal(ast):
```
Comment (ErebusZ) [inline src/ostorlab/serve_app/common.py:75]: same here

File: src/ostorlab/serve_app/app.py [hunk 15]
```diff
@@ -32,3 +36,79 @@ def authenticate() -> Optional[tuple[flask.Response, int]]:
...
+        if content_type == "application/ubjson":
+            return self.ubjson_decode(flask.request)
+        return super().parse_body()
+
+    def ubjson_decode(self, request: flask.Request):
+        """Decode UBJSON request data."""
+
+        return ubjson.loadb(request.data)
+
+    def dispatch_request(self):
```
Comment (ErebusZ) [inline src/ostorlab/serve_app/app.py:57]: typing.

**Comment by 3asm** [comment]
@elyousfi5 what is the status of this PR since it is failing?

**Review by 3asm — APPROVED** [review]
